text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,157 @@
+
+When a new topic is registered through the `casper_register_message_topic` FFI, the host creates a composite key derived from the callers entity address and the topic name.
+This step is required in order for users to be able to verify messages later on by querying global state using a `Key` derived as mentioned above. This also allows for specifying a tighter limit on the number of messages that can be generated.
+
+When the host receives these messages through the `casper_emit_message` FFI, the identity of the caller entity will be attached to the message.
+The messages that are emitted will not be stored in global state; only a digest of the identity and the message will be however stored to global state. This allows users to verify the integrity of the message and also check that the message was indeed generated by the expected entity by doing a query to global state.
+Messages will be passed by the execution engine to the node that will eventually send them out on the event stream after the execution is complete and committed.
+
+**Note:** the messages themselves cannot be reconstructed by reading the entry from global state since that entry contains only the digest of the messages sent.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation for supporting contract level messages requires changes in both the execution engine and the node.
+
+### Changes to execution engine
+
+#### Introduce new FFIs:
+```rust
+/// Registers a new message topic with the specified name for the calling contract.
+/// 
+/// The topic name is a string of a maximum length that is specified in the chainspec.
+/// Will return an error if:
+///     * the topic name length is exceeded
+///     * the topic size is larger than the maximum topic size specified in the chainspec
+pub fn casper_register_message_topic(topic_name_ptr: *const u8, topic_name_size: usize, topic_size: usize)
+```
+
+```rust
+/// Emits a new message on the specified topic.
+/// 
+/// The topic needs to have been already registered through a call to `casper_register_message_topic` before emitting a message.
+/// 
+/// Will return an error if:
+///     * the maximum number of messages on this topic were exceeded
+///     * the message is of type [`Message`]
+///     * the length of the message is larger than the maximum length specified in the chainspec
+pub fn casper_emit_message(
+    topic_name_ptr: *const u8,
+    topic_name_size: usize,
+    message_ptr: *const u8,
+    message_size: usize,
+);
+```
+
+The `Message` enum initially allows for sending human readable strings as messages in the first iteration:
+```rust
+pub enum Message {
+    String(String)
+}
+```
+
+#### A new `Key` variant is introduced for indexing the message topic digests:
```

Comment:
First, thank you for the excellent CEP.

I would suggest a tweak to the architecture. Rather than include execution effects in global storage, I would recommend we construct a separate Merkle tree of deploy execution effects and include a Merkle root in each block (following Ethereum's architecture for transaction receipts). We would not have to persist this Merkle tree, we would compute the Merkle root for it and only persist canonical deploy execution effects (indexed by deploy) in storage.

Does that make sense?

Reply:
Thank you for the feedback! Your proposal is interesting.
However going through some experiments it became apparent that we would probably want to allow users to query for checksums of distinct messages.
With the previous design the application listening for messages on a topic would have needed to catch all messages to be able to verify the whole topic which is a bit unrealistic.

So with this new requirement I think we would have to also persist this new Merkle tree that you are proposing as well to be able to query for any discrete message checksum. Not sure if there's any gain in having it separate rather than just writing to global state in this case.

I've done some modifications to this CEP in the latest commit so please take a look; the changes are mainly related to how we validate these discrete messages and added a bit more info on gas cost and limits.

Reply:
> So with this new requirement I think we would have to also persist this new Merkle tree that you are proposing as well to be able to query for any discrete message checksum. Not sure if there's any gain in having it separate rather than just writing to global state in this case.

You would not need to persist this Merkle tree -- Ethereum does not persist Merkle trees for its contract level events.  You just compute the Merkle root and store the events.  There is already an implementation of this in the `casper-node` repo: https://github.com/casper-network/casper-node/blob/29466f9089d309720e2d8515a3b649a1e0a4fbae/hashing/src/lib.rs#L156

In addition, have you looked at [Ethereum's bloom filter design](https://ethereum.github.io/execution-specs/autoapi/ethereum/frontier/bloom/index.html) for its events?  This is commonly used by light clients as it avoids the need to download all contract logs when checking a block.

Reply:
There's also code for computing a Merkle proof of inclusion given a Merkle root: https://github.com/casper-network/casper-node/blob/29466f9089d309720e2d8515a3b649a1e0a4fbae/hashing/src/indexed_merkle_proof.rs#L63

Reply:
> > So with this new requirement I think we would have to also persist this new Merkle tree that you are proposing as well to be able to query for any discrete message checksum. Not sure if there's any gain in having it separate rather than just writing to global state in this case.
> 
> You would not need to persist this Merkle tree -- Ethereum does not persist Merkle trees for its contract level events. You just compute the Merkle root and store the events. There is already an implementation of this in the `casper-node` repo: https://github.com/casper-network/casper-node/blob/29466f9089d309720e2d8515a3b649a1e0a4fbae/hashing/src/lib.rs#L156

We wouldn't want to store the full events though. For this to provide an advantage we would probably want to pass on the merkle proof along with the events on the SSE to allow consumers to verify it. But most likely we would want to store some information (e.g. the number of messages that were emitted on each topic in order to see if a particular type of message was emitted at all). Given this I think we get some additional advantage if we store this type of info in Global State as proposed in the CEP: since consumers can listen for the execution journal, they can more accurately reconstruct the context in which that message was emitted should they require to.
This gives a bit more transparency to the system rather than providing only proofs for the events emitted that are rather opaque WRT context.